cps/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayFlowFactoryAction.java

```diff
@@ -0,0 +1,103 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.cps.replay;
+
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.InvisibleAction;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowFactoryAction2;
+import org.jenkinsci.plugins.workflow.cps.steps.LoadStepExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+
+/**
+ * Attached to a run that is a replay of an earlier one.
+ */
+class ReplayFlowFactoryAction extends InvisibleAction implements CpsFlowFactoryAction2, Queue.QueueAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ReplayFlowFactoryAction.class.getName());
+
+    private String mainScript;
+    private final Map<String,String> otherScripts;
+    private transient final boolean sandbox;
+    
+    ReplayFlowFactoryAction(@Nonnull String mainScript, @Nonnull Map<String,String> otherScripts, boolean sandbox) {
+        this.mainScript = mainScript;
+        this.otherScripts = new HashMap<String,String>(otherScripts);
+        this.sandbox = sandbox;
+    }
+
+    @Override public CpsFlowExecution create(FlowDefinition def, FlowExecutionOwner owner, List<? extends Action> actions) throws IOException {
+        String script = mainScript;
+        mainScript = null; // minimize build.xml size
+        return new CpsFlowExecution(script, sandbox, owner);
+    }
+
+    @Override public boolean shouldSchedule(List<Action> actions) {
+        return true; // do not coalesce
+    }
+
+    @Extension public static class ReplacerImpl implements LoadStepExecution.Replacer {
```

Comment:
Thanks, that helps; I think I misunderstood the intent here; but I'm still a little confused what the intent is for multiple replacements here?  Does it make sense to have multiple extension point impls, any one of which defines the final returned text?  We've marked the ExtensionPoint as NoExternalUse, is the for loop here mostly just boilerplate and something that will need to change if we allow for other extensions anyway? 

I suppose what I'm getting at here is wondering if the API allows for future extension as it stands.